### ISSUE TITLE
Fetching data in Distributed Tensorflow has too much latency 

### ISSUE THREAD

**Issue by ahaider3** [tensorflow/tensorflow#11411]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:

https://gist.github.com/ahaider3/ae4f6d2d790d963a93b346bb0138a41d

The above is a simple benchmark which tests the overhead of distributed TF. It fetches a configurable sized variable from the parameter server and does a matmul on the worker. It also does a matmul from a locally stored variable on the worker. The time difference between these two operations would be the overhead I am measuring. 

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Linux Redhat  
- **TensorFlow installed from (source or binary)**:
source
- **TensorFlow version (use command below)**:
tensorflow 1.2 
- **Python version**: 
python 2.7.7

- **Exact command to reproduce**:

python matmul_benchmark.py --num_features=256 --batch_size=128 --num_hidden=64 --job_name=ps &

python matmul_benchmark.py --num_features=256 --batch_size=128 --num_hidden=64 --job_name=worker

By increasing batch_size, the timing difference between local/remote computation eventually becomes negligible. 
However, for small batch sizes the overhead can become 2x/3x:

For example, here are two runs for different model/dataset sizes:

> 128 features, batch size of 32, hidden layer size of 256  returns:
Local GEMM Time:  0.0002624  Network Fetch GEMM Time: 0.0006798

> 256 features, batch size of 128, hidden layer size of 128  returns:
Local GEMM Time: 0.0002995 Network Fetch GEMM Time: 0.0006124

### Describe the problem
Distributed tensorflow introduces overhead due to its communication stack. By overhead I mean the additional time required for workers to receive data from parameter servers when compared to doing the same computation without fetching any remote data. 

This is a problem because due to this overhead I have to use 2/3 nodes to just provide performance on-par with non-distributed (single process) tensorflow. The number of nodes required to be on-par with single process TF increases further when I use gpus. 

Fetching small variables provides a constant overhead which limits scaling and efficiency . 
This overhead creates two issues in Distributed Tensorflow:
1) I have to add several workers just to equal the performance of a single process. 
2) The overhead of fetching model parameters doesn't scale but the amount of computation does 
decrease as I add more workers. Thus, once I get to a moderately small batch size for each worker I can't scale because the constant overhead of fetching remote model parameters. 

There have been several issues posted with distributed tensorflow. #6116 is an improvement to large tensor transfer while this problem exists for small tensors. #4498 might have been caused by CPU performance bottleneck and not network. However for my problem, network transfer is definitely the bottleneck. I have tried using RDMA and have seen minimal benefit. 

### Source code / logs
https://gist.github.com/ahaider3/ae4f6d2d790d963a93b346bb0138a41d

**Comment by byronyi** [comment]
This is a known issue (#4498, #6116, #11196) with several workarounds:

1. Wait until #7466 (cause:  #6116) is merged, which gives roughly 2-3x speedup to gRPC large tensor transport.
2. Use alternative communication protocol, such as `grpc+verbs` or `grpc+mpi`. This requires RDMA capable hardware.
3. Refactor your code and try to use intra-process communication.

**Comment by ahaider3** [comment]
I am mostly interested in the performance of small tensors. I have tried grpc+verbs and haven't seen any significant benefits. I will try with MPI.

**Comment by byronyi** [comment]
I see both of your model size and batch size is small (128/256?). What's the time spent on computation each round before fetching model parameters from PS? If it only spends tens to hundreds of microseconds, I don't think the current distributed runtime could reach a performance on par to that of a single process, as there's a fixed overhead on setting up interprocess communication.

**Comment by ahaider3** [comment]
Without fetching model parameters it only takes a few hundred microseconds. Yes, that's the problem I have. My compute time per iteration is low and so going to distributed requires double/triple the time per iteration due to the fixed overhead. 

I didn't see much benefit when I was testing with grpc+verbs, but that was with larger tensors. Is there a difference between `grpc+mpi` or `grpc+verbs` when compared to `grpc` besides the actual communication protocol? Are tensors still serialized in the same manner. The actual communication across the network is not the bottleneck from my testing.

**Comment by byronyi** [comment]
If you are in doubt, you could try my own [GPU Direct patch](#11392) which is theoretically of lowest latency. It does neither memory copy nor serialization. Btw if your computation is not that complicated, you should avoid using GPU; copying data from/to GPU adds nontrivial overhead in your communication pipeline.

**Comment by eamartin** [comment]
I'm also interested in this issue, with the application of training small feedforward nets (say 200-100-50-4) as quickly as possible. Note this only has ~100KB of parameters. I'd like to run on an algorithmic minibatch size of 8K, split up among as many workers as possible (so 8 workers means each handles 1K examples per iteration). As I add more workers, the batch size per worker gets smaller, which makes connection overhead more significant (per Amdahl's law).

I'm working on training these on CPU, storing the variables on a parameter server. I've found the connection overhead destroys parallelism, as this issue points out. The particular problem here is not low throughput for transferring large parameters, it is high latency in moving even very small amounts of data. My guess is this comes from connection setup time, and that steps such as serialization and extra copies do not help. 

I have not tried ``gprc+verbs`` or ``grpc+mpi``, and I don't expect them to help significantly as my understanding is these protocols do the same connection establishment and then just move the tensors themselves (tiny amounts of data in my case!) through fast networking. Is this a valid understanding? Is there anywhere (docs or code) that I can learn more about network protocol involved in fetching tensors from parameter server.

I implemented the same model with data parallelism using MPI (instead of TensorFlow's networking stack) to allreduce the gradients and found hugely better performance (running at 2 nodes results in a ~1.8x speedup instead of a slowdown).

@byronyi Is their any way to avoid tensor serialization on CPU?

**Comment by byronyi** [comment]
@eamartin In my patch, tensors are transferred in terms of direct memory access (DMA) of the underlying tensor buffer, so there is no serialization nor memory copy. It is titled "GPU Direct RDMA", but certainly works for CPU as well.

**Comment by jart** [comment]
Thanks for helping our friend @byronyi.

**Comment by eamartin** [comment]
Can we pause on closing this until the mentioned patch is actually merged? We have not yet confirmed that serialization or memcpy's caused the overhead. Other culprits could be multiple network roundtrips causing extra latency (and killing bandwidth in the small tensor limit).

**Comment by byronyi** [comment]
Take a look at #10530 and you may find something that interests you. 

Let me know if you have further questions; I am more than willing to help you out.

**Comment by suiyuan2009** [comment]
I run the script provided by@yaroslavvb in #4498, result for `grpc` is 
```
Local rate:       15962.31 MB per second
Distributed rate: 335.68 MB per second
```
and `grpc+verbs` is
```
Local rate:       15514.35 MB per second
Distributed rate: 1306.15 MB per second
```
I'm using a 56Gbps ib network.

**Comment by byronyi** [comment]
I just ran in a 40Gbps RoCE setup for my GDR patch, and here's the result:

Adding data in 100 MB chunks
Local rate: 5243.48 MB per second
Distributed rate: 2679.18 MB per second

Numbers for `grpc` and `grpc+verbs` are similar to what you got (~300 MB/s and ~1300MB/s).

**Comment by ahaider3** [comment]
@suiyuan2009 I am not seeing that benefit from `grpc+rdma` on my ib network. Are your results with an updated patch? I am running with TF 1.2.0.

Here are tests with different chunk sizes:

For 100 MB chunks with grpc:

> Local rate:       19761.24 MB per second
   Distributed rate: 339.71 MB per second


For 512 KB **chunks with grpc:

> Local rate:       3538.90 MB per second
  Distributed rate: 616.97 MB per second


So decreasing the chunk size by `200x` only increases throughput by `2x`. I think the problem of large tensors is important, but small tensor transfer is also slow.

**Comment by jart** [comment]
@byronyi Please be sure to add `Fixes #11411` to your commit message in #11392.

**Comment by suiyuan2009** [comment]
@ahaider3 , I built from official master branch, I'll try @byronyi 's branch. I find there is not much difference between benchmark scripts which run distributed tensorflow on same machine or different machines. The script in #4498 performs bettern on different machines than on a single machine, weird.

**Comment by suiyuan2009** [comment]
I find that when test with small data(10MB for example), `grpc+rdma` 's performance is very bad, speed decreases from `1300MB/s` to less than `1000MB/s`(300MB/s or 900MB/s, not stable).

**Comment by byronyi** [comment]
@ahaider3 @suiyuan2009 There is an important patch #10531 got merged after the 1.2 release, so it is expected that the current master is faster than 1.2 for `grpc+verbs` runtime.

**Comment by suiyuan2009** [comment]
I find `assign` is much slow than `assign_add`, but I think `assign` should be as fast as `assign_add` at least.

**Comment by byronyi** [comment]
@suiyuan2009 

I've noticed this issue, too. I'm testing my GDR patch using the [benchmark_grpc_recv.py](https://gist.github.com/yaroslavvb/e196107b5e0afc834652bd3153030c42) script (courtesy @yaroslavvb).

For `assign`, 3 measurements in a row:

```
Local rate:       6944.22 MB/s
Distributed rate: 2690.64 MB/s
---
Local rate:       5084.81 MB/s
Distributed rate: 2910.57 MB/s
---
Local rate:       5864.24 MB/s
Distributed rate: 2588.69 MB/s
```

For `assign_add`:

```
Local rate:       16558.85 MB/s
Distributed rate: 3248.83 MB/s
---
Local rate:       9952.02 MB/s
Distributed rate: 3681.21 MB/s
---
Local rate:       16090.50 MB/s
Distributed rate: 3418.83 MB/s
```

But the variance of these measurements (running on the same machine) seems to be just as large as the throughput gap. So I would rather not take it as a serious issue or a performance bug.

**Comment by byronyi** [comment]
@eamartin As GDR is available in current master #11392, would you mind to try again with the newer `grpc+gdr` protocol? It does no tensor serialisation nor memory copies for tensor data.

Since all three extension protocol (`grpc+gdr`, `grpc+verbs` and `grpc+mpi`) still use grpc heavily for the control plane, @ahaider3 you might not see much difference as your computation is way too fast compared to the fixed overhead of setting up each tensor transmission (it is rather a latency issue, not a throughput one). I personally would try to port grpc from HTTP2 to RDMA, but it will be a patch unlikely to be accepted by the grpc project as indicated [here](https://github.com/tensorflow/tensorflow/issues/2916#issuecomment-253924638).

@jart Any comments?

**Comment by futurely** [comment]
https://github.com/caffe2/caffe2/tree/master/caffe2/contrib/gloo
https://github.com/facebookincubator/gloo#benchmarking

**Comment by ahaider3** [comment]
I ran the tests again using TF 1.2 with the different extension protocols that I can run on my system. In my case, I can only use `grpc+verbs`.

I found minimal to no benefit from using this protocol for my small model. I am measuring the throughput of fetching a parameter and then doing a GEMM. 

> 128 features; 32 batch size; 256 hidden layer size
grpc:  192.81 MB/sec
grpc+verbs: 179.18 MB/sec

This was the average of three runs. I would conclude that these protocols are limited by GRPC because of the latency it introduces, as suggested by @byronyi . 

There has been interesting work by Uber to make communication all-around more efficient: https://github.com/uber/horovod

@jart what do you think about uber's method for distributed training.

**Comment by byronyi** [comment]
Horovod uses NCCL 2, which supports InfiniBand but not RoCE. See [here](https://github.com/uber/horovod/issues/5).

**Comment by littleDing** [comment]
I report the same problem as @ahaider3 that small mini-batch on small model is terrible for distributed training using cpu.

When I train a deep&wide model with about 15 feas, embedding_size=32, I tried the 2/3/4 machine of ps+worker and ends up with the same speed as one machine of all in one process. 
Only after increasing embedding_size to 256 give a 2x speed up with 4 machine.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @jart: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by byronyi** [comment]
I think we should close this issue as it appears to be resolved?

**Comment by jart** [comment]
Thanks for the tip. Closing now that PR is merged. https://github.com/tensorflow/tensorflow/pull/11392
